### PR TITLE
In waitForDeploymentStatus, list RSes once rather than twice

### LINKED ISSUE #22719: e2e flake: deployment should support rollover [minUnavailable/maxSurge violated]
kubernetes-jenkins/logs/kubernetes-e2e-gce/12967
```
• Failure [314.132 seconds]
Deployment
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:72
  deployment should support rollover [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:59

  Expected error:
      <*errors.errorString | 0xc2084d9e90>: {
          s: "failed to wait for pods running: [gave up waiting for pod 'test-rollover-controller-0pi42' to be 'running' after 5m0s]",
      }
      failed to wait for pods running: [gave up waiting for pod 'test-rollover-controller-0pi42' to be 'running' after 5m0s]
  not to have occurred

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:461
```
cc @janetkuo

### PR DESCRIPTION
**Pull request by janetkuo**
Fixes #22719

@kubernetes/sig-config

### CODE REVIEW

File: test/e2e/util.go [hunk 1]
```diff
@@ -2465,11 +2465,7 @@ func waitForDeploymentStatus(c clientset.Interface, ns, deploymentName string, d
 		if err != nil {
 			return false, err
 		}
-		oldRSs, allOldRSs, err = deploymentutil.GetOldReplicaSets(deployment, c)
```
Comment (bgrant0607) [inline test/e2e/util.go]: Good catch. Yes, these were mutating operations.

**Comment by bgrant0607** [comment]
How about waitForDeploymentRevisionAndImage?

**Comment by bgrant0607** [comment]
And waitForDeploymentOldRSsNum?

File: pkg/util/deployment/deployment.go [hunk 2]
```diff
@@ -48,6 +48,39 @@ const (
...
+	for _, rs := range rsList.Items {
+		if err != nil {
+			return nil, nil, fmt.Errorf("invalid label selector: %v", err)
+		}
+		// Filter out replica set that has the same pod template spec as the deployment - that is the new replica set.
+		if api.Semantic.DeepEqual(rs.Spec.Template, newRSTemplate) {
+			newRS = &rs
+			continue
+		}
+		oldRSes[rs.ObjectMeta.Name] = rs
```
Comment (bgrant0607) [inline pkg/util/deployment/deployment.go]: There shouldn't be multiple ReplicaSets with the same name in the same namespace. The system guarantees spatial uniqueness.

File: pkg/util/deployment/deployment.go [hunk 3]
```diff
@@ -48,6 +48,39 @@ const (
 	RollbackDone              = "DeploymentRollback"
 )
 
+// GetAllReplicaSets list all RSes of a given deployment atomically
+func GetAllReplicaSets(deployment *extensions.Deployment, c clientset.Interface) (allOldRSes []*extensions.ReplicaSet, newRS *extensions.ReplicaSet, err error) {
```
Comment (kargakis) [inline pkg/util/deployment/deployment.go]: Why can't we re-use this helper for all list operations in the deployment controller? Do we guarantee that the rs will have as labels the deployment selector? Related issue: https://github.com/kubernetes/kubernetes/issues/17491

File: pkg/util/deployment/deployment.go (hunk 2 above)
Comment (kargakis) [inline pkg/util/deployment/deployment.go]: Agreed. Taking into account https://github.com/kubernetes/kubernetes/pull/22889/files#r55919190 and that this opeartion is popular (listing replicasets in the deployment controller), we may want to drop the map at some point and just append to the slice.

File: pkg/util/deployment/deployment.go (hunk 3 above)
Comment (bgrant0607) [inline pkg/util/deployment/deployment.go]: The deployment controller adds labels matching the selector to the ReplicaSets, yes.

**Comment by bgrant0607** [comment]
From in-person discussion: I don't think we need to solve the consistency problem right now. The test doesn't appear to be flaking that often. However, I do think we need the e2e tests to not mutate any of the resources unintentionally.

**Comment by janetkuo** [comment]
> I do think we need the e2e tests to not mutate any of the resources unintentionally.

#22948 to solve this

**Comment by bgrant0607** [comment]
Removing this issue from 1.2, in favor of #22948

**Comment by bgrant0607** [comment]
This PR now just does one RS list call instead of 2 separate lists?

**Comment by bgrant0607** [comment]
e2e failed:
```
â€¢ Failure [325.115 seconds]
Deployment
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:79
  deployment should support rollback when there's replica set with no revision [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:75

  Expected error:
      <*errors.errorString | 0xc2084f3290>: {
          s: "error waiting for deployment test-rollback-no-revision-deployment status to match expectation: timed out waiting for the condition",
      }
      error waiting for deployment test-rollback-no-revision-deployment status to match expectation: timed out waiting for the condition
  not to have occurred

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:772
```

**Comment by bgrant0607** [comment]
@janetkuo Do we still need/want this?

**Comment by janetkuo** [comment]
This will mitigate but does not solve the consistency problem. Closing this.

**Pull request closed without merging**